### DISCUSSION TITLE
Implementing Torque Offset Modifications in the ROSCO Controller Source Code

### DISCUSSION THREAD

**Discussion by dorsamgh** [NatLabRockies/ROSCO#405] [Announcements]
Hello, I hope you are doing well. I am a beginner in working with the ROSCO controller. However, I successfully compiled the standard version of it using the Anaconda command environment. 

My goal is to implement a time-based offset condition in the wind turbine torque. To achieve this, I referred to the `Controllers.f90` source code and attempted to make modifications in the `VariableSpeedControl` subroutine to incorporate this feature. Below, I have highlighted the changes I made:
![image](https://github.com/user-attachments/assets/9f1b1797-8572-4380-b613-4ff29a90c155)

Despite saving the changes, recompiling, generating the new DLL file, and running the final simulation, I observed no difference compared to the normal state. 

I suspect the issue might be related to the `VS_ControlMode` variable. Despite thoroughly searching through the source codes, I couldn't identify where this parameter is set. 

I would greatly appreciate it if you could also guide me on this matter.
I would appreciate it if you could guide me on this issue. 

I am using ROSCO version 2.9.0, OpenFAST version 3.3.0, and the IEA-15-240-RWT-UMaineSemi model. Additionally, the wind speed in the `bts` file is set to 9 m/s, which falls within Region 2.

Thank you in advance for your help.

**Comment by dorsamgh** [comment]
I have now realized that based on the DISCON input file, the value of VS_ControlMode is set to 2. Therefore, I need to make this change in this section:
![image](https://github.com/user-attachments/assets/904a8c8e-01f2-423a-8f92-73b7534cd695)


I have implemented this change and plotted the torque difference compared to the normal mode in a graph. I expected this difference to be 100 kNm from second 400 until the end of the simulation. However, this has not occurred as expected. Could you please guide me regarding this issue?
![image](https://github.com/user-attachments/assets/ce6e1ec8-f52b-4c95-b4e9-7995a1600001)

**Comment by dzalkind** [comment]
Hi,

Since the PI torque controller is trying to regulate the speed to the desired set point, the torque will adapt to the offset you have provided and find a new steady state torque (offset by the opposite of your value) to regulate the speed to the desired set point.

You have two options:
1. Change the rotor speed set point.  On the `develop` branch, there is the capability to change the power reference in time using `R_Speed`.  Below rated, this will change the rotor speed set point by the factor `R_Speed`.  https://github.com/NREL/ROSCO/blob/develop/Examples/29_power_control.py
2. Use `VS_ControlMode` of 1, which will use the k\omega^2 law to control the power when not in the transition regions.  This is still a form of closed loop control, so by changing the torque, you'll change the speed and thus the torque again.  
 
Option 1 is probably your best bet, even though you're not directly changing the torque offset.

I hope this helps.  I'm not sure there is an issue with the code, so I may move this to a discussion in the near future.  For now, I'll keep it here.

Best, Dan
